pkg/kubelet/kuberuntime/fake_kuberuntime.go

```diff
@@ -0,0 +1,554 @@
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "RemoveContainer")
+
+	if _, ok := r.Containers[rawContainerID]; ok {
+		delete(r.Containers, rawContainerID)
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) ListContainers(filter *runtimeApi.ContainerFilter) ([]*runtimeApi.Container, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "ListContainers")
+
+	result := make([]*runtimeApi.Container, 0)
+	for _, s := range r.Containers {
+		if filter != nil {
+			if filter.Id != nil && filter.GetId() != s.containerID {
+				continue
+			}
+			if filter.Name != nil && filter.GetName() != s.containerConfig.GetName() {
+				continue
+			}
+			if filter.PodSandboxId != nil && filter.GetPodSandboxId() != s.podSandBoxID {
+				continue
+			}
+			if filter.State != nil && filter.GetState() != s.state {
+				continue
+			}
+			if filter.LabelSelector != nil && !filterInLabels(filter.LabelSelector, s.containerConfig.Labels) {
+				continue
+			}
+		}
+
+		result = append(result, &runtimeApi.Container{
+			Id:       &s.containerID,
+			Name:     s.containerConfig.Name,
+			State:    &s.state,
+			Image:    s.containerConfig.Image,
+			ImageRef: &s.imageID,
+			Labels:   s.containerConfig.Labels,
+		})
+	}
+
+	return result, nil
+}
+
+func (r *fakeKubeRuntime) ContainerStatus(rawContainerID string) (*runtimeApi.ContainerStatus, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "ContainerStatus")
+
+	c, ok := r.Containers[rawContainerID]
+	if !ok {
+		return nil, fmt.Errorf("Not found")
+	}
+
+	return &runtimeApi.ContainerStatus{
+		Id:        &c.containerID,
+		Name:      c.containerConfig.Name,
+		State:     &c.state,
+		CreatedAt: &c.createdAt,
+		Image:     c.containerConfig.Image,
+		ImageRef:  &c.imageID,
+		Labels:    c.containerConfig.Labels,
+	}, nil
+}
+
+func (r *fakeKubeRuntime) Exec(rawContainerID string, cmd []string, tty bool, stdin io.Reader, stdout, stderr io.WriteCloser) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "Exec")
+	return nil
+}
+
+// getContainerByKubeUID gets a list of ContainerWithState by pod UID
+func (r *fakeKubeRuntime) getContainersByKubeUID(uid string) []*ContainerWithState {
+	result := []*ContainerWithState{}
+
+	for _, c := range r.Containers {
+		if c.containerConfig.Labels[types.KubernetesPodUIDLabel] == uid {
+			result = append(result, c)
+		}
+	}
+
+	return result
+}
+
+// getPodSandboxByKubeUID gets a list of PodSandboxWithState by pod UID
+// It returns a list because there may be more than one pod sandboxes belonging to one Pod (e.g. stopped)
+func (r *fakeKubeRuntime) getPodSandboxByKubeUID(uid string) []*PodSandboxWithState {
+	result := []*PodSandboxWithState{}
+
+	for _, s := range r.Sandboxes {
+		if s.config.Labels[types.KubernetesPodUIDLabel] == uid {
+			result = append(result, s)
+		}
+	}
+
+	return result
+}
+
+// getExpectedKubePodStatus get kubecontainer.PodStatus by api.Pod, should only be used after syncPod
```

Comment:
We should deal with the excessive types later (`kubecontainer.Container`, `runtimeApi.Container`, etc).